File: Example/AFView/Common/AFKeyboardInfo.m
#import "AFKeyboardInfo.h"


#pragma mark Class Definition

@implementation AFKeyboardInfo


#pragma mark - Properties

- (UIViewAnimationOptions)animationCurveAsOptions
{
	return UNAnimationOptionsFromAnimationCurve(_animationCurve);
}


#pragma mark - Constructors

- (id)initWithEndFrame: (CGRect)endFrame
	animationDuration: (NSTimeInterval)animationDuration
	animationCurve: (UIViewAnimationCurve)animationCurve
	originalOptions: (NSDictionary *)originalOptions
{
	// Abort if base initializer fails.
	if ((self = [super init]) == nil)
	{
		return nil;
	}
	
	// Apply a minimum animation duration.
	if (animationDuration == 0)
	{
		animationDuration = 0.2f;
	}
	
	// Initialize instance variables.
	_endFrame = endFrame;
	_animationCurve = animationCurve;
	_originalOptions = originalOptions;
	_animationDuration = animationDuration;
	
	// Return initialized instance.
	return self;
}


#pragma mark - Methods

- (void)animate: (void (^)(void))animations
{
	// Animate using the keyboard options animation options.
	[UIView beginAnimations: nil
		context: NULL];
	[UIView setAnimationDuration: [_originalOptions[UIKeyboardAnimationDurationUserInfoKey] doubleValue]];
	[UIView setAnimationCurve: [_originalOptions[UIKeyboardAnimationCurveUserInfoKey] integerValue]];
	[UIView setAnimationBeginsFromCurrentState: YES];

	// Can be anything.
	if (animations != nil)
	{
		animations();
	}
	
	// Commit the animations.
	[UIView commitAnimations];
}

static inline UIViewAnimationOptions UNAnimationOptionsFromAnimationCurve(UIViewAnimationCurve animationCurve)
{	
	switch (animationCurve)
	{
		case UIViewAnimationCurveEaseInOut:
			return UIViewAnimationOptionCurveEaseInOut;
		case UIViewAnimationCurveEaseIn:
			return UIViewAnimationOptionCurveEaseIn;
		case UIViewAnimationCurveEaseOut:
			return UIViewAnimationOptionCurveEaseOut;
		case UIViewAnimationCurveLinear:
			return UIViewAnimationOptionCurveLinear;
	}
	
	return 0;
}


@end

File: AFView/Pod/Common/AFKeyboardInfo.swift
import Foundation

public class AFKeyboardInfo : NSObject
{
	// MARK: - Properties

	public var animationCurveAsOptions: UIViewAnimationOptions
	{
		return AFKeyboardInfo.animationOptionsFromAnimationCurve(self.animationCurve)
	}

	var animationCurve: UIViewAnimationCurve
	var animationDuration: NSTimeInterval
	var originalOptions: NSDictionary
	var endFrame: CGRect


	// MARK: - Constructors

	public init(endFrame: CGRect,
		animationDuration: NSTimeInterval,
		animationCurve: UIViewAnimationCurve,
		originalOptions: NSDictionary)
	{
		// Apply a minimum animation duration.
		if (animationDuration == 0)
		{
			self.animationDuration = 0.2
		}
		else
		{
			self.animationDuration = animationDuration
		}
	
		// Initialize instance variables.
		self.endFrame = endFrame
		self.animationCurve = animationCurve
		self.originalOptions = originalOptions
		
		super.init()
	}


	// MARK: - Public Methods

	public func animate(animations: (() -> ())?)
	{
		// Animate using the keyboard options animation options.
		UIView.beginAnimations(nil,
			context: nil)
		
		// Set the duration.
		if let rawDuration = self.originalOptions[UIKeyboardAnimationDurationUserInfoKey]
		{
			UIView.setAnimationDuration(rawDuration.doubleValue)
		}
		
		// Set the curve.
		if let rawCurve = self.originalOptions[UIKeyboardAnimationCurveUserInfoKey]
		{
			UIView.setAnimationCurve(UIViewAnimationCurve(rawValue: rawCurve.integerValue)!)
		}
		
		UIView.setAnimationBeginsFromCurrentState(true)
		
		// Can be anything.
		if let animations = animations
		{
			animations();
		}
		
		// Commit the animations.
		UIView.commitAnimations()
	}
	
	public static func animationOptionsFromAnimationCurve(animationCurve: UIViewAnimationCurve)
		-> UIViewAnimationOptions
	{
		switch (animationCurve)
		{
			case .EaseInOut:
				return UIViewAnimationOptions.CurveEaseInOut
			case .EaseIn:
				return UIViewAnimationOptions.CurveEaseIn
			case .EaseOut:
				return UIViewAnimationOptions.CurveEaseOut
			case .Linear:
				return UIViewAnimationOptions.CurveLinear
		}
	}
}